UserSessionsMeter.c

```diff
@@ -0,0 +1,162 @@
+(C) 2024 htop dev team
+Released under the GNU GPLv2+, see the COPYING file
+in the source distribution for its full text.
+*/
+
+#include "config.h" // IWYU pragma: keep
+
+#include "UserSessionsMeter.h"
+
+#include <dlfcn.h>
+#include <utmpx.h>
+
+
+#include "CRT.h"
+#include "Macros.h"
+#include "Object.h"
+#include "RichString.h"
+#include "XUtils.h"
+
+#if defined(BUILD_STATIC) && defined(HAVE_LIBSYSTEMD)
+#include <systemd/sd-login.h>
+#endif
+
+
+static unsigned int users = 0;
+
+#ifdef BUILD_STATIC
+#define sym_sd_get_sessions sd_get_sessions
+#else
+
+static int (*sym_sd_get_sessions)(char***);
+static void* dlopenHandle = NULL;
+
+#endif /* BUILD_STATIC */
+
+
+static void UserSessionsMeter_done(ATTR_UNUSED Meter* this) {
+#ifndef BUILD_STATIC
+   if (dlopenHandle) {
+      dlclose(dlopenHandle);
+      dlopenHandle = NULL;
+   }
+#endif /* !BUILD_STATIC */
+}
+
+#if !defined(BUILD_STATIC) || defined(HAVE_LIBSYSTEMD)
+static int update_via_sd_login(void) {
+#ifndef BUILD_STATIC
+   if (!dlopenHandle) {
+      dlopenHandle = dlopen("libsystemd.so.0", RTLD_LAZY);
+      if (!dlopenHandle)
+         goto dlfailure;
+
+      /* Clear any errors */
+      dlerror();
+
+      #define resolve(symbolname) do {                                      \
+         *(void **)(&sym_##symbolname) = dlsym(dlopenHandle, #symbolname);  \
+         if (!sym_##symbolname || dlerror() != NULL)                        \
+            goto dlfailure;                                                 \
+      } while(0)
+
+      resolve(sd_get_sessions);
+
+      #undef resolve
+   }
+#endif /* !BUILD_STATIC */
+
+   return sym_sd_get_sessions(NULL);
+
+#ifndef BUILD_STATIC
+dlfailure:
+   if (dlopenHandle) {
+      dlclose(dlopenHandle);
+      dlopenHandle = NULL;
+   }
+   return -1;
+#endif /* !BUILD_STATIC */
+}
+#endif /* !BUILD_STATIC || HAVE_LIBSYSTEMD */
+
+static int updateViaUtmp(void) {
+   struct utmpx *ut;
+   int ret = 0;
+#ifdef HAVE_STRUCT_UTMPX_UT_SESSION
+   long sessions[32];
+
+   for (size_t i = 0; i < ARRAYSIZE(sessions); i++)
+      sessions[i] = -1;
+#endif /* HAVE_STRUCT_UTMPX_UT_SESSION */
+
+   setutxent();
+   while ((ut = getutxent())) {
+      if (ut->ut_type == USER_PROCESS) {
+#ifdef HAVE_STRUCT_UTMPX_UT_SESSION
+         for (size_t i = 0; i < ARRAYSIZE(sessions); i++) {
+            if (sessions[i] == ut->ut_session) {
+               break;
+            }
+
+            if (sessions[i] == -1) {
+               ret++;
+               sessions[i] = ut->ut_session;
+               break;
+            }
+         }
+#else
+      ret++;
+#endif /* HAVE_STRUCT_UTMPX_UT_SESSION */
+      }
+   }
+   endutxent();
+
+   return ret;
+}
+
+static void UserSessionsMeter_updateValues(Meter* this) {
+   int ret;
+#if !defined(BUILD_STATIC) || defined(HAVE_LIBSYSTEMD)
+   ret = update_via_sd_login();
+   if (ret < 0)
+      ret = updateViaUtmp();
+#else
+   ret = updateViaUtmp();
+#endif /* !BUILD_STATIC || HAVE_LIBSYSTEMD */
```

Comment:
1. Can the systemd code depend only on `HAVE_LIBSYSTEMD`? People who build htop without systemd support would likely not want the systemd code even when it's dynamically loaded and linked.
2. Code simplification suggestion:


```c
   int ret = -1; // Or maybe -ENOSYS for better diagnostic message
#if defined(HAVE_LIBSYSTEMD)
   ret = update_via_sd_login();
#endif /* HAVE_LIBSYSTEMD */
   if (ret < 0)
      ret = updateViaUtmp();
```

Reply:
1. There is a systemd meter. Linking should be done accordingly as it's already done there.

2. This looks a bit cleaner, but has a dead store in `ret`.